### PR TITLE
Issue 147 - Derived virtual/viewport selected rows

### LINKED ISSUE #147: Add derived properties `derived_virtual_selected_rows` and `derived_viewport_selected_rows`
`selected_rows` doesn't update on sort or filter

See this example: https://github.com/plotly/dash-docs/pull/232/files#diff-61a5d030fd2c372f3abe94110f039e85

### PR DESCRIPTION
**Pull request by Marc-Andre-Rivet**
Fixed #147.

Adds support for two new derived readonly props:
- (selected_rows, virtual_indices) --> `derived_virtual_selected_rows`
- (selected_rows, viewport_indices) --> `derived_viewport_selected_rows`

The indices used in the derived props are the virtual and viewport ones, not the original data indices. For example, with

df_indices = [0...10[
df_virtual_indices = [0, 2, 4, 6, 8] (filter even rows)
df_viewport_indices = [0, 2, 4] (page size = 3)

selected_rows = [0, 2, 6, 7]
derived_virtual_selected_rows = [0, 1, 3]
derived_viewport_selected_rows = [0, 1]

Need to add a server or jest/enzyme test to make sure the derived props are updated as expected.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "dash-table",
-  "version": "3.1.3",
+  "version": "3.1.4",
```
Comment (Marc-Andre-Rivet) [inline package.json]: bump version

File: src/dash-table/components/Table/index.tsx [hunk 2]
```diff
@@ -81,6 +83,16 @@ export default class Table extends Component<PropsWithDefaultsAndDerived> {
...
 
+        const virtual_selected_rows = this.virtualSelectedRows(
+            virtual.indices,
+            selected_rows
+        );
+
+        const viewport_selected_rows = this.viewportSelectedRows(
+            viewport.indices,
+            selected_rows
+        );
```
Comment (Marc-Andre-Rivet) [inline src/dash-table/components/Table/index.tsx]: The calculation for both virtual and viewport derived props is the same, only the data changes

File: src/dash-table/components/Table/index.tsx [hunk 3]
```diff
@@ -134,6 +162,14 @@ export default class Table extends Component<PropsWithDefaultsAndDerived> {
             newProps.derived_viewport_indices = viewport.indices;
         }
 
+        if (!virtualSelectedRowsCached) {
+            newProps.derived_virtual_selected_rows = virtual_selected_rows;
+        }
+
+        if (!viewportSelectedRowsCached) {
+            newProps.derived_viewport_selected_rows = viewport_selected_rows;
+        }
```
Comment (Marc-Andre-Rivet) [inline src/dash-table/components/Table/index.tsx]: Update if the derived prop changed

File: src/dash-table/components/Table/index.tsx [hunk 4]
```diff
@@ -162,12 +198,16 @@ export default class Table extends Component<PropsWithDefaultsAndDerived> {
 
     private readonly paginator = derivedPaginator();
     private readonly viewport = derivedViewportData();
+    private readonly viewportSelectedRows = derivedSelectedRows();
     private readonly virtual = derivedVirtualData();
+    private readonly virtualSelectedRows = derivedSelectedRows();
```
Comment (Marc-Andre-Rivet) [inline src/dash-table/components/Table/index.tsx]: Using the same function but two different memoization.

File: src/dash-table/components/Table/index.tsx [hunk 5]
```diff
@@ -162,12 +198,16 @@ export default class Table extends Component<PropsWithDefaultsAndDerived> {
...
     private readonly visibleColumns = derivedVisibleColumns();
 
     private readonly filterCache = memoizeOneWithFlag(filter => filter);
     private readonly paginationCache = memoizeOneWithFlag(pagination => pagination);
     private readonly sortCache = memoizeOneWithFlag(sort => sort);
     private readonly viewportCache = memoizeOneWithFlag(viewport => viewport);
+    private readonly viewportSelectedRowsCache = memoizeOneWithFlag(viewport => viewport);
     private readonly virtualCache = memoizeOneWithFlag(virtual => virtual);
+    private readonly virtualSelectedRowsCache = memoizeOneWithFlag(virtual => virtual);
```
Comment (Marc-Andre-Rivet) [inline src/dash-table/components/Table/index.tsx]: Each with its own cache busting "instance"

File: src/dash-table/components/Table/props.ts [hunk 6]
```diff
@@ -130,7 +130,7 @@ interface IProps {
     row_deletable?: boolean;
     row_selectable?: RowSelection;
     selected_cells?: SelectedCells;
-    selected_rows?: number[];
+    selected_rows?: Indices;
```
Comment (Marc-Andre-Rivet) [inline src/dash-table/components/Table/props.ts]: Improving typing to be consistent with other usages

File: src/dash-table/components/Table/props.ts [hunk 7]
```diff
@@ -198,8 +198,10 @@ interface IDefaultProps {
 interface IDerivedProps {
     derived_viewport_data: Data;
     derived_viewport_indices: Indices;
+    derived_viewport_selected_rows: Indices;
     derived_virtual_data: Data;
     derived_virtual_indices: Indices;
+    derived_virtual_selected_rows: Indices;
```
Comment (Marc-Andre-Rivet) [inline src/dash-table/components/Table/props.ts]: The new derived props

File: src/dash-table/derived/selectedRows/index.ts [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+
+const getter = (
+    indices: Indices,
+    selectedRows: Indices
+): Indices => {
+    const map = new Map<number, number>();
+
+    R.addIndex<number>(R.forEach)((virtualIndex, index) => {
+        map.set(virtualIndex, index);
+    }, indices);
```
Comment (Marc-Andre-Rivet) [inline src/dash-table/derived/selectedRows/index.ts]: Map each original df index to its position in the derived indices.

File: src/dash-table/derived/selectedRows/index.ts [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+
+    const mappedSelectedRows: number[] = [];
+    R.forEach(rowIndex => {
+        const virtualIndex = map.get(rowIndex);
+
+        if (virtualIndex !== undefined) {
+            mappedSelectedRows.push(virtualIndex);
+        }
+    }, selectedRows);
```
Comment (Marc-Andre-Rivet) [inline src/dash-table/derived/selectedRows/index.ts]: Prune the selected that do not have a matching entry in the derived indices

File: src/dash-table/DataTable.js [hunk 10]
```diff
@@ -717,7 +725,13 @@ export const propTypes = {
...
     derived_virtual_indices: PropTypes.arrayOf(PropTypes.number),
 
-     /**
+    /**
+     * `derived_virtual_selected_rows` represents the indices of the
+     *  `selected_rows` from the perspective of the `derived_virtual_indices`.
+     */
+    derived_virtual_selected_rows: PropTypes.arrayOf(PropTypes.number),
+
+    /**
```
Comment (Marc-Andre-Rivet) [inline src/dash-table/DataTable.js]: two new props exposed by the table

File: src/dash-table/components/Table/index.tsx [hunk 11]
```diff
@@ -134,12 +158,24 @@ export default class Table extends Component<PropsWithDefaultsAndDerived> {
...
+
         if (invalidateSelection) {
             newProps.active_cell = undefined;
             newProps.selected_cells = undefined;
             newProps.selected_rows = undefined;
         }
 
+        if (!R.keys(newProps).length) {
+            return;
+        }
```
Comment (Marc-Andre-Rivet) [inline src/dash-table/components/Table/index.tsx]: If there are no props in newProps, don't bother firing a setProps..

File: tests/cypress/tests/server/select_row_test.ts [hunk 12]
```diff
@@ -22,4 +23,56 @@ describe('select row', () => {
             DashTable.getSelect(0).within(() => cy.get('input').should('not.be.checked'));
         });
     });
+
+    describe('fe pagination & sort', () => {
```
Comment (Marc-Andre-Rivet) [inline tests/cypress/tests/server/select_row_test.ts]: New tests for derived selected rows props -- check that the Dash Server receives the correct values for derived_virtual_selected_rows and derived_viewport_selected_rows with/without sort and filtering

File: tests/cypress/tests/unit/derivedSelectedRows_test.ts [hunk 13]
```diff
@@ -0,0 +1,133 @@
+import * as R from 'ramda';
+
+import derivedSelectedRows from 'dash-table/derived/selectedRows';
+
+describe('derived selected rows', () => {
```
Comment (Marc-Andre-Rivet) [inline tests/cypress/tests/unit/derivedSelectedRows_test.ts]: Unit tests on the derived selected rows calculation

**Comment by chriddyp** [comment]
Does the example in the docs work if you modify it to use `derived_virtual_selected_rows`? That is, this example: https://dash.plot.ly/datatable/interactivity

**Comment by Marc-Andre-Rivet** [comment]
@chriddyp Yes, it seems to be working well.

issue147

This might be a Dash thing, but I had to add this check because they initial value of `derived_virtual_selected_rows` is None even if the default is `[]`

```
    if selected_rows is None:
        selected_rows = []
```

**Comment by chriddyp** [comment]
> This might be a Dash thing, but I had to add this check because they initial value of derived_virtual_selected_rows is None even if the default is []

Yeah, that's a Dash thing 😬 

> Yes, it seems to be working well.

Awesome, thanks for checking it out. 💃 from me for the API.

**Comment by Marc-Andre-Rivet** [comment]
@chriddyp Will publish as 3.1.4 and update dash-docs

**Pull request merged**